// @angular-component/router/esm5/lib/router.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { LocationStrategy, PathLocationStrategy, CommonModule, } from '@angular/common';
import { RouterComponent } from './router.component';
import { RouteComponent } from './route.component';
import { RouteComponentTemplate } from './route-component.directive';
import { LinkActive } from './link-active.directive';
import { LinkTo } from './link-to.directive';
import { UrlParser } from './url-parser';
import { QueryParams } from './route-params.service';
import { Router } from './router.service';
export var components = [
    RouterComponent,
    RouteComponent,
    LinkActive,
    LinkTo,
    RouteComponentTemplate,
];
export function getQueryParams(router) {
    return router.queryParams$;
}
var ComponentRouterModule = /** @class */ (function () {
    function ComponentRouterModule() {
    }
    ComponentRouterModule_1 = ComponentRouterModule;
    ComponentRouterModule.forRoot = function () {
        return {
            ngModule: ComponentRouterModule_1,
            providers: [
                UrlParser,
                { provide: LocationStrategy, useClass: PathLocationStrategy },
                { provide: QueryParams, deps: [Router], useFactory: getQueryParams },
            ],
        };
    };
    var ComponentRouterModule_1;
    ComponentRouterModule = ComponentRouterModule_1 = __decorate([
        NgModule({
            imports: [CommonModule],
            declarations: [components],
            exports: [components],
        })
    ], ComponentRouterModule);
    return ComponentRouterModule;
}());
export { ComponentRouterModule };
